community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/BadCollector.java

```diff
@@ -29,11 +29,23 @@
 
 import static java.lang.String.format;
 import static java.util.Arrays.copyOf;
+import static java.util.Arrays.sort;
 
 import static org.neo4j.helpers.Exceptions.withMessage;
 
 public class BadCollector implements Collector
 {
+    /**
+     * Introduced to avoid creating an exception for every reported bad thing, since it can be
+     * quite the performance hogger for scenarios where there are many many bad things to collect.
+     */
+    private interface Bad
+    {
+        String message();
+
+        InputException exception();
+    }
```

Comment:
Perhaps `ProblemReporter` or `DeferredInputProblem` are (slightly) better names? "Bad" isn't very descriptive. I'm also not super excited about the `BadCollector` name, but can't think of something better right now.